File: packages/emitter/testcases/real-world/shapes/shapes.cs
namespace TestCases.realworld.shapes
{
    public class Shape
    {
        public string color;

        public Shape(string color)
            {
            this.color = color;
            }

        public double area();

        public double perimeter();

        public string describe()
            {
            return $"A {this.color} shape with area {this.area()} and perimeter {this.perimeter()}";
            }
    }
    public class Circle : Shape
    {
        public double radius;

        // ERROR: super() must be the first statement in constructor
        public Circle(string color, double radius)
        {
            // Constructor body omitted due to error
        }

        public override double area()
            {
            return global::Tsonic.JSRuntime.Math.PI * this.radius * this.radius;
            }

        public override double perimeter()
            {
            return 2 * global::Tsonic.JSRuntime.Math.PI * this.radius;
            }
    }
    public class Rectangle : Shape
    {
        public double width;

        public double height;

        // ERROR: super() must be the first statement in constructor
        public Rectangle(string color, double width, double height)
        {
            // Constructor body omitted due to error
        }

        public override double area()
            {
            return this.width * this.height;
            }

        public override double perimeter()
            {
            return 2 * this.width + this.height;
            }

        public bool isSquare()
            {
            return this.width == this.height;
            }
    }
    public class Triangle : Shape
    {
        public double @base;

        public double height;

        public double side1;

        public double side2;

        // ERROR: super() must be the first statement in constructor
        public Triangle(string color, double @base, double height, double side1, double side2)
        {
            // Constructor body omitted due to error
        }

        public override double area()
            {
            return this.@base * this.height / 2;
            }

        public override double perimeter()
            {
            return this.@base + this.side1 + this.side2;
            }
    }

            public static class shapes
            {
                public static double totalArea(global::System.Collections.Generic.List<Shape> shapes)
                    {
                    return global::Tsonic.JSRuntime.Array.reduce(shapes, (sum, shape) => sum + shape.area(), 0);
                    }

                public static Shape? findLargestShape(global::System.Collections.Generic.List<Shape> shapes)
                    {
                    if (global::Tsonic.Runtime.Array.length(shapes) == 0)
                        {
                        return default;
                        }
                    var largest = global::Tsonic.Runtime.Array.get(shapes, 0);
                    foreach (var shape in shapes)
                        {
                        if (shape.area() > largest.area())
                            {
                            largest = shape;
                            }
                        }
                    return largest;
                    }
            }
}

File: packages/emitter/testcases/real-world/shapes/shapes.ts
export abstract class Shape {
  constructor(public color: string) {}

  abstract area(): number;
  abstract perimeter(): number;

  describe(): string {
    return `A ${this.color} shape with area ${this.area()} and perimeter ${this.perimeter()}`;
  }
}

export class Circle extends Shape {
  constructor(
    color: string,
    public radius: number
  ) {
    super(color);
  }

  area(): number {
    return Math.PI * this.radius * this.radius;
  }

  perimeter(): number {
    return 2 * Math.PI * this.radius;
  }
}

export class Rectangle extends Shape {
  constructor(
    color: string,
    public width: number,
    public height: number
  ) {
    super(color);
  }

  area(): number {
    return this.width * this.height;
  }

  perimeter(): number {
    return 2 * (this.width + this.height);
  }

  isSquare(): boolean {
    return this.width === this.height;
  }
}

export class Triangle extends Shape {
  constructor(
    color: string,
    public base: number,
    public height: number,
    public side1: number,
    public side2: number
  ) {
    super(color);
  }

  area(): number {
    return (this.base * this.height) / 2;
  }

  perimeter(): number {
    return this.base + this.side1 + this.side2;
  }
}

export function totalArea(shapes: Shape[]): number {
  return shapes.reduce(
    (sum: number, shape: Shape): number => sum + shape.area(),
    0
  );
}

export function findLargestShape(shapes: Shape[]): Shape | undefined {
  if (shapes.length === 0) {
    return undefined;
  }

  let largest = shapes[0];
  for (const shape of shapes) {
    if (shape.area() > largest.area()) {
      largest = shape;
    }
  }
  return largest;
}
